<div style="border: 1px solid thistle; padding: 5px 20px 10px 20px;">
    <h2><font color = "77658f" style=";">Feedback</font></h2><br>
    <font color = "77658f"> • Good work for EDA and data cleaning with the removal all duplicates before modelling <br>• Elaborate more to explain your working through the use of more markdowns  <br>• The model may work bettwe with TFIDF Vectorizer.  Consider removing subreddit topics and also review parameters used for gridsearch especially the use of even bigrams and trigrams<br>• Do also examine the misclassified posts more closely for similar tokens <br><br>
        • Problem Statement: <b>2</b> <br>
        • Data Collection: <b>2</b> <br>
        • Data Cleaning and EDA: <b>2</b> <br>
        • Preprocessing and Modeling: <b>2</b><br>
        • Evaluation and Conceptual Understanding: <b>2</b><br>
        • Conclusion and Recommendation: <b>2</b><br>
        • Organization: <b>2</b><br>
        • Visualizations: <b>2</b><br>
        • Python Syntax and Control Flow: <b>2</b><br>
        • Presentation: <b>3</b><br>
    </font>
</div>

# Project 3: Part I

- Problem Statement
- Executive Summary
- Data Collection
- Simple Cleaning of Data


> <font size = 3 color = "crimson"> Do create links to the section headers   </font>

# Problem Statement

Discussions on Reddit are organized into user-created areas of interest called "subreddits". There are about 138,000 active subreddits among a total of 1.2 million, as of July 2018. As the amount of subreddits will continue to grow, we would like build a classifier model to help identify posts from different subbredits and guide user traffic to the appropiate subbredit.

> <font size = 3 color = "crimson"> Include evaluation metric. Noted that impact to stakeholders is in executive summary   </font>

# Executive Summary:
The reason to create this classifier model is so that we can help administrators to summarise a whole thread, suppress redundant or irrelevant posts. For contributers, it might provide guidance for increasing the usefulness of a contribution. For moderators, it might to act as a base model to identify things such as spam posts, misplaced posts, flame wars, troll users or threads that need locking, merging or deleting.

In [1]:
# import the libraries
import requests
import pandas as pd
import time
import random

from bs4 import BeautifulSoup

# Data Collection

I picked the below 2 subbreddits on romance and relationships as they are both similar in topics and to see if the classifier is capable of differentiating between these 2 subreddits. 

In [2]:
# targetted urls for webscraping

url1 = "https://www.reddit.com/r/romance.json"
url2 = "https://www.reddit.com/r/relationships.json"

I will scrape 1000 posts from reddits API for each subbreddit

> <font size = 3 color = "crimson"> Consider writing a function for the code below with your topics as arguments.</font>

In [20]:
# Webscrape r/romance subreddit

posts = [] # list to store the posts
after = None

for a in range(40):
    if after == None:
        current_url = url1
    else:
        current_url = url1 + '?after=' + after
    print(current_url)
    res = requests.get(current_url, headers={'User-agent': 'JP Inc 1.0'})
    
    if res.status_code != 200:
        print('Status error', res.status_code)
        break
        
    # Read in posts for the current page
    current_dict = res.json()
    current_posts = [p['data'] for p in current_dict['data']['children']]
    print("no. of post: " + str(len(current_posts)))
    posts.extend(current_posts)
    after = current_dict['data']['after']
    
    # if more than 1 page, read previous pages then merge with current page, write back to csv file
    if a > 0:
        prev_posts = pd.read_csv('romance.csv')
        current_df = pd.DataFrame(posts)
        combined_posts = pd.concat([prev_posts,current_df], ignore_index=True)
        combined_posts.to_csv('romance.csv', index = False)
        
    else:
        pd.DataFrame(posts).to_csv('romance.csv', index = False)

    # generate a random sleep duration to look more 'natural'
    sleep_duration = random.randint(2,6)
    print(sleep_duration)
    time.sleep(sleep_duration)

https://www.reddit.com/r/romance.json
no. of post: 25
2
https://www.reddit.com/r/romance.json?after=t3_gctr16
no. of post: 25
5
https://www.reddit.com/r/romance.json?after=t3_g4lkuz
no. of post: 25
6
https://www.reddit.com/r/romance.json?after=t3_fszgsb
no. of post: 25
5
https://www.reddit.com/r/romance.json?after=t3_flmf7c
no. of post: 25
2
https://www.reddit.com/r/romance.json?after=t3_fdqqhn
no. of post: 25
6
https://www.reddit.com/r/romance.json?after=t3_f7niat
no. of post: 25
5
https://www.reddit.com/r/romance.json?after=t3_f0494b
no. of post: 25
4
https://www.reddit.com/r/romance.json?after=t3_eslf7q
no. of post: 25
6
https://www.reddit.com/r/romance.json?after=t3_emcbqx
no. of post: 25
4
https://www.reddit.com/r/romance.json?after=t3_efpj5g
no. of post: 25
3
https://www.reddit.com/r/romance.json?after=t3_e7zbqa
no. of post: 25
2
https://www.reddit.com/r/romance.json?after=t3_e05nxt
no. of post: 25
3
https://www.reddit.com/r/romance.json?after=t3_ds95n2
no. of post: 25
5
https://

In [7]:
# Webscrape r/relationships subreddit

posts = []
after = None

for a in range(40):
    if after == None:
        current_url = url2
    else:
        current_url = url2 + '?after=' + after
    print(current_url)
    res = requests.get(current_url, headers={'User-agent': 'JP Inc 1.0'})
    
    if res.status_code != 200:
        print('Status error', res.status_code)
        break
        
    # Read in posts for the current page
    current_dict = res.json()
    current_posts = [p['data'] for p in current_dict['data']['children']]
    print("no. of post: " + str(len(current_posts)))
    posts.extend(current_posts)
    after = current_dict['data']['after']
    
    # if more than 1 page, read previous pages then merge with current page, write back to csv file
    if a > 0:
        prev_posts = pd.read_csv('relationships.csv')
        current_df = pd.DataFrame(posts)
        combined_posts = pd.concat([prev_posts,current_df], ignore_index=True)
        combined_posts.to_csv('relationships.csv', index = False)
        
    else:
        pd.DataFrame(posts).to_csv('relationships.csv', index = False)

    # generate a random sleep duration to look more 'natural'
    sleep_duration = random.randint(2,6)
    print(sleep_duration)
    time.sleep(sleep_duration)

https://www.reddit.com/r/relationships.json
no. of post: 26
4
https://www.reddit.com/r/relationships.json?after=t3_gi0s8d
no. of post: 25
4
https://www.reddit.com/r/relationships.json?after=t3_ghyzlo
no. of post: 25
6
https://www.reddit.com/r/relationships.json?after=t3_ghyp0t
no. of post: 25
6
https://www.reddit.com/r/relationships.json?after=t3_gi60e9
no. of post: 25
4
https://www.reddit.com/r/relationships.json?after=t3_gi3xu3
no. of post: 25
2
https://www.reddit.com/r/relationships.json?after=t3_ghuy77
no. of post: 25
4
https://www.reddit.com/r/relationships.json?after=t3_ghyur2
no. of post: 25
4
https://www.reddit.com/r/relationships.json?after=t3_ghwxud
no. of post: 25
4
https://www.reddit.com/r/relationships.json?after=t3_ghny0z
no. of post: 25
5
https://www.reddit.com/r/relationships.json?after=t3_ghsen0
no. of post: 25
3
https://www.reddit.com/r/relationships.json?after=t3_ghpzg2
no. of post: 25
5
https://www.reddit.com/r/relationships.json?after=t3_ghm62q
no. of post: 25
3
ht

## Simple Cleaning of data
I will now proceed to check the data and to remove any null or duplicate posts.

In [36]:
rom_df = pd.read_csv('romance.csv')
rel_df = pd.read_csv('relationships.csv')

In [37]:
rom_df.shape

(993, 104)

In [38]:
rel_df.shape

(979, 103)

Seems like there are some missing data in relationships dataframe or the posts have reached their limit. There also seem to be 1 missing column in relationship dataframe compared to romance dataframe. This should not be an issue as long as that missing column is not the column containing our posts. 

In [39]:
# Display to see dataframe max columns of 104
pd.options.display.max_columns = 104

In [40]:
rom_df.head()

approved_at_utc subreddit  \
0              NaN   romance   
1              NaN   romance   
2              NaN   romance   
3              NaN   romance   
4              NaN   romance   

                                            selftext author_fullname  saved  \
0  I am 21 years old and I’ve never been in love....     t2_6eot1ted  False   
1  I usually don't do this but my heart beats to ...      t2_nxyxdb2  False   
2  I know marriage is not exactly a necessity and...      t2_nowm36j  False   
3  I like you— more than your bravery and intelli...     t2_6eoh35vg  False   
4  I could usually go out of my way and flirt or ...     t2_4e8rqi12  False   

   mod_reason_title  gilded  clicked  \
0               NaN       0    False   
1               NaN       0    False   
2               NaN       0    False   
3               NaN       0    False   
4               NaN       0    False   

                                               title link_flair_richtext  \
0                                  I Desire Romance.                  []   
1  Going through a tough break up right now so de...                  []   
2  It’s my biggest dream to get married and I hop...                  []   
3                                  To someone I like                  []   
4  Is it actually possible to fall in love with a...                  []   

  subreddit_name_prefixed  hidden  pwls  link_flair_css_class  downs  \
0               r/romance   False   NaN                   NaN      0   
1               r/romance   False   NaN                   NaN      0   
2               r/romance   False   NaN                   NaN      0   
3               r/romance   False   NaN                   NaN      0   
4               r/romance   False   NaN                   NaN      0   

   hide_score       name  quarantine link_flair_text_color  upvote_ratio  \
0       False  t3_ghiik1       False                  dark          0.81   
1       False  t3_ghfzef       False                  dark          1.00   
2       False  t3_gh1biw       False                  dark          1.00   
3       False  t3_ggxdcj       False                  dark          1.00   
4       False  t3_gh3r13       False                  dark          1.00   

   author_flair_background_color subreddit_type  ups  total_awards_received  \
0                            NaN         public    3                      0   
1                            NaN         public    3                      0   
2                            NaN         public   16                      0   
3                            NaN         public    7                      0   
4                            NaN         public    1                      0   

  media_embed  author_flair_template_id  is_original_content user_reports  \
0          {}                       NaN                False           []   
1          {}                       NaN                False           []   
2          {}                       NaN                False           []   
3          {}                       NaN                False           []   
4          {}                       NaN                False           []   

   secure_media  is_reddit_media_domain  is_meta  category secure_media_embed  \
0           NaN                   False    False       NaN                 {}   
1           NaN                   False    False       NaN                 {}   
2           NaN                   False    False       NaN                 {}   
3           NaN                   False    False       NaN                 {}   
4           NaN                   False    False       NaN                 {}   

   link_flair_text  can_mod_post  score  approved_by author_premium  \
0              NaN         False      3          NaN          False   
1              NaN         False      3          NaN          False   
2              NaN         False     16          NaN          False   
3              NaN         False      

In [52]:
rel_df.head()

approved_at_utc      subreddit  \
0              NaN  relationships   
1              NaN  relationships   
2              NaN  relationships   
3              NaN  relationships   
4              NaN  relationships   

                                            selftext author_fullname  saved  \
0  Dear /r/relationships community,\n\nWe recogni...        t2_ahi9l  False   
1  I'm very openly bisexual, and my friend "Lily"...     t2_6fzpu1aw  False   
2  Link to the original post:\n\n [https://www.re...     t2_4d6cmw7i  False   
3  We’ve been together for 3 years. Recently he’s...     t2_4zxugwnn  False   
4  So this literally just happened. My husband wa...     t2_4qfxw6b7  False   

   mod_reason_title  gilded  clicked  \
0               NaN       0    False   
1               NaN       0    False   
2               NaN       0    False   
3               NaN       0    False   
4               NaN       0    False   

                                               title link_flair_richtext  \
0  Coronavirus (SARS-CoV2, COVID-19) Mega-thread ...                  []   
1  My [21F] best friend [20F]'s bf [30M] won't le...                  []   
2  UPDATE: Coworker (M) is almost certainly recor...                  []   
3  He [22M] wouldn’t let me [21F] see his phone a...                  []   
4  My (28F) husband (30M) just walked out. What now?                  []   

  subreddit_name_prefixed  hidden  pwls link_flair_css_class  downs  \
0         r/relationships   False     6            meta meta      0   
1         r/relationships   False     6            m-gn misc      0   
2         r/relationships   False     6         m-it updates      0   
3         r/relationships   False     6        m-be breakups      0   
4         r/relationships   False     6   m-vt relationships      0   

   hide_score       name  quarantine link_flair_text_color  upvote_ratio  \
0       False  t3_g79ye3       False                  dark          0.94   
1       False  t3_ghzarl       False                  dark          0.97   
2       False  t3_ghnm3z       False                  dark          0.99   
3       False  t3_gi7fuk       False                  dark          0.89   
4       False  t3_gi2awb       False                  dark          0.91   

   author_flair_background_color subreddit_type   ups  total_awards_received  \
0                            NaN         public    85                      0   
1                            NaN         public  2180                      0   
2                            NaN         public  3794                      0   
3                            NaN         public    40                      0   
4                            NaN         public   104                      0   

  media_embed  author_flair_template_id  is_original_content user_reports  \
0          {}                       NaN                False           []   
1          {}                       NaN                False           []   
2          {}                       NaN                False           []   
3          {}                       NaN                False           []   
4          {}                       NaN                False           []   

   secure_media  is_reddit_media_domain  is_meta  category secure_media_embed  \
0           NaN                   False    False       NaN                 {}   
1           NaN                   False    False       NaN                 {}   
2           NaN                   False    False       NaN                 {}   
3           NaN                   False    False       NaN                 {}   
4           NaN                   False    False       NaN                 {}   

  link_flair_text  can_mod_post  score  approved_by  author_premium  \
0        « Meta »         False     85          NaN            True   
1    Non-Romantic         False   2180          NaN           False   
2         Updates         False   3794          NaN           False   
3       

In [86]:
# check if string
rom_df['selftext'].dtype

dtype('O')

In [87]:
# check if string
rel_df['selftext'].dtype

dtype('O')

Columns that I would be interested to work on would be the main content of the posts which are in the 'selftext' column and I will keep the 'subreddit' column as a label to identify where my post originate from.  

In [41]:
# Check for null columns
rom_df['selftext'].isnull().sum()

68

In [42]:
# Check for null columns
rom_df['subreddit'].isnull().sum()

0

In [43]:
# Check for null columns
rel_df['selftext'].isnull().sum()

0

In [44]:
# Check for null columns
rel_df['subreddit'].isnull().sum()

0

Seems like there are missing data in rom_df['selftext']. I will drop those rows.

In [45]:
# Drop null rows
rom_df = rom_df.dropna(subset=['selftext', 'subreddit'])

In [46]:
rom_df.shape

(925, 104)

In [47]:
# Drop any duplicate entries
rom_df = rom_df.drop_duplicates(subset=['selftext'])

In [48]:
rom_df.shape

(924, 104)

In [49]:
rel_df.shape

(979, 103)

In [19]:
rel_df['selftext'].value_counts().sort_values(ascending=False)

My husband is an only child and I just have the one sister. With all of this coronavirus hoopla I thought it would be best to try to square away our will. My dad passed away and it's just my mom. My husband's parents are a bit older and not in the best of health to care for a child.\n\nI've gotten closer to my sister in the past few years as growing up we were too far apart in age.\n\nI thought she would be the perfect candidate for a guardian as she seems to adore my son and my husband gets along with her as well. We have life insurance and had plans for the rest of our assets.\n\nWhen we asked her to be the guardian we told her that the only caveat would be if we are ok with her future husband. (I would tell her first and foremost if I thought he was abusive or not good for her, this wasn't about if we could all go vacationing or something). She was immediately ecstatic and said yes, and we were glad.\n\nShe called a few days ago stating that she wasn't ready for a commitment and she

In [50]:
# Drop any duplicate entries
rel_df = rel_df.drop_duplicates(subset=['selftext'])

In [51]:
rel_df.shape

(506, 103)

For rom_df, there is a 1 duplicate entry which was dropped, leaving me with 924 posts.<br>
For rel_df, there are 473 posts which were dropped, leaving me with 506 posts.

I will randomly sample 506 posts out of 924 posts from rom_df as I would like have a balance sample size between my catergorical data.

In [54]:
rom_samp_df = rom_df.sample(n=506, random_state=42)

In [55]:
rom_samp_df.shape

(506, 104)

I will write my clean data sets into new .csv files for processing

In [56]:
clean_rom_df = pd.DataFrame(rom_samp_df, columns=['selftext', 'subreddit'])

In [57]:
clean_rom_df.shape

(506, 2)

In [58]:
# export to new .csv
pd.DataFrame(clean_rom_df).to_csv('clean_romance.csv', index = False)

In [52]:
clean_rel_df = pd.DataFrame(rel_df, columns=['selftext', 'subreddit'])

In [53]:
clean_rel_df.shape

(506, 2)

In [59]:
# export to new .csv
pd.DataFrame(clean_rel_df).to_csv('clean_relationships.csv', index = False)

EDA, model building and predictions will be carried out in the next jupyter notebook